src/librustc/traits/error_reporting.rs

```diff
@@ -837,6 +833,50 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
         err
     }
+
+    fn report_closure_arg_mismatch(&self,
+                           span: Span,
+                           found_span: Option<Span>,
+                           expected_ref: ty::PolyTraitRef<'tcx>,
+                           found: ty::PolyTraitRef<'tcx>)
+        -> DiagnosticBuilder<'tcx>
+    {
+        fn build_args_type_string(span: Span, trait_ref: &ty::TraitRef) -> String {
+            match trait_ref.substs.type_at(1).sty {
+                ty::TyTuple(args, _) => {
+                    let len = args.len();
+                    assert_ne!(len, 0);
+
+                    let mut type_list_string = String::new();
+                    if len >= 2 {
+                        for ty in args[..len-2].iter() {
+                            type_list_string.push_str(&format!("`{}`, ", ty));
+                        }
+                        type_list_string.push_str(&format!("`{}` and ", args[len-2]));
+                    }
+                    type_list_string.push_str(&format!("`{}`", args[len-1]));
+                    format!("argument{} of type {}",
+                            if len == 1 { "" } else { "s" },
+                            type_list_string)
+                },
+                _ => span_bug!(span, "generic type parameter is not a tuple"),
+            }
+        }
+
+        let mut err = struct_span_err!(self.tcx.sess, span, E0631,
+                                       "type mismatch in closure arguments");
+        let found_arg_list = found.map_bound_ref(
+            move |trait_ref| build_args_type_string(span, trait_ref)
+        );
+        err.span_label(span, format!("expected {}", found_arg_list.skip_binder()));
+        if let Some(sp) = found_span {
```

Comment:
You should display the "takes" even when you don't have a span - aka fn pointers:
```Rust
fn foo(f: fn((u32, u32))) {
    a.iter().map(f)
}
```

Reply:
I think you should add a test for each case with fn pointer - arg count mismatch, same arg count but type-error, higher-ranked error, F: Fn<u32> bad bound.

Reply:
Higher-ranked test for fn pointers:
```Rust
fn foo<F: Fn(*mut &u32)>(_f: F) {}
fn bar<'a>(f: fn(*mut &'a u32)) { foo(f); }
```

Higher-ranked test for closures: src/test/ui/mismatched_types/closure-mismatch.rs (already exists)

Just check that the error message is sane.

Reply:
Can I use `note` here?

Reply:
With respect to the span specifically, probably just something like this would be fine:

```
let found_span = found_span.unwrap_or(span);
```

i.e., if we don't have a better span, display both messages at the same place.